test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+      })
+    })
+  })
+
+  describe("permit", () => {
+    const permittingHolderBalance = to1e18(650000)
+    let permittingHolder
+
+    let yesterday
+    let tomorrow
+
+    beforeEach(async () => {
+      permittingHolder = await ethers.Wallet.createRandom()
+      await token.mint(permittingHolder.address, permittingHolderBalance)
+
+      const lastBlockTimestamp = await lastBlockTime()
+      yesterday = lastBlockTimestamp - 86400 // -1 day
+      tomorrow = lastBlockTimestamp + 86400 // +1 day
+    })
+
+    const getApproval = async (amount, spender, deadline) => {
+      // We use ethers.utils.SigningKey for a Wallet instead of
+      // Signer.signMessage to do not add '\x19Ethereum Signed Message:\n'
+      // prefix to the signed message. The '\x19` protection (see EIP191 for
+      // more details on '\x19' rationale and format) is already included in
+      // EIP2612 permit signed message and '\x19Ethereum Signed Message:\n'
+      // should not be used there.
+      const signingKey = new ethers.utils.SigningKey(
+        permittingHolder.privateKey
+      )
+
+      const domainSeparator = await token.DOMAIN_SEPARATOR()
+      const permitTypehash = await token.PERMIT_TYPEHASH()
+      const nonce = await token.nonces(permittingHolder.address)
+
+      const approvalDigest = ethers.utils.keccak256(
+        ethers.utils.solidityPack(
+          ["bytes1", "bytes1", "bytes32", "bytes32"],
+          [
+            "0x19",
+            "0x01",
+            domainSeparator,
+            ethers.utils.keccak256(
+              ethers.utils.defaultAbiCoder.encode(
+                [
+                  "bytes32",
+                  "address",
+                  "address",
+                  "uint256",
+                  "uint256",
+                  "uint256",
+                ],
+                [
+                  permitTypehash,
+                  permittingHolder.address,
+                  spender,
+                  amount,
+                  nonce,
+                  deadline,
+                ]
+              )
+            ),
+          ]
+        )
+      )
+
+      return ethers.utils.splitSignature(
+        await signingKey.signDigest(approvalDigest)
+      )
+    }
+
+    context("when permission expired", () => {
+      it("should revert", async () => {
+        const deadline = yesterday
+        const signature = await getApproval(
+          permittingHolderBalance,
+          anotherAccount.address,
+          deadline
+        )
+
+        await expect(
+          token
+            .connect(anotherAccount)
+            .permit(
+              permittingHolder.address,
+              anotherAccount.address,
+              permittingHolderBalance,
+              deadline,
+              signature.v,
+              signature.r,
+              signature.s
+            )
+        ).to.be.revertedWith("Permission expired")
+      })
+    })
+
+    context("when permission has an invalid signature", () => {
```

Comment:
We should cover two revert cases:
- `Invalid signature 's' value`
- `Invalid signature 'v' value`